File: py/lib/blockchain.py
"""Blockchain in Python."""

import os
import sys


# pylint: disable=C0411
from .block import Block
from .blockdb import BlockDB, packint
from .crypto import Hasher, ECDSAPrivateKey
from .querylayer import QueryLayer
from .transaction import Transaction, TransactionInput, TransactionOutput
from .transdb import TransDB

TOTAL_COINS = 987654321000


class BlockChain:
    """Blockchain object."""

    def __init__(self, base_dir: str) -> None:
        """Create a new BlockChain object."""
        self.base_dir = base_dir
        self.blockdb = BlockDB(base_dir=self.base_dir)
        self.transdb = TransDB(base_dir=self.base_dir)
        self.q = QueryLayer(bdb=self.blockdb, tdb=self.transdb)
        self.blockdb.consistency_check(self.q)
        return

    def create(self, passphrase: str) -> None:
        """Create a new blockchain (must not already exist!)."""
        assert self.blockdb.num_blocks == 0, "Blockchain already exists!"

        print("Creating new blockchain...")
        # Create initial keys.
        pk = ECDSAPrivateKey.generate()
        try:
            key_prefix = os.path.join(self.base_dir, "genesis_key")
            pk.write_key_pair(filename_prefix=key_prefix,
                              passphrase=passphrase)
        except FileExistsError as exc:
            print("Unable to write key file: {}".format(exc))
            print("Blockchain not created.")
            sys.exit(1)

        # Add genesis block manually.
        block = Block(0, Hasher(b"There can be only one").get_hash())
        address = pk.publickey.get_address()
        to = TransactionOutput(address=address, amount=TOTAL_COINS)

        # Fake the UTXO in.
        txid = Hasher(b"Genesis").get_hash()
        utxokey = self.blockdb.get_key_for_utxo(txid, address)
        self.blockdb.db.put(utxokey, packint(TOTAL_COINS))

        # Sign the input.
        output_hash = Hasher([
            to.get_hash()
        ]).get_hash()
        sig = Hasher([
            txid,
            output_hash
        ]).sign(pk)

        ti = TransactionInput(txid=txid, pubkey=pk.publickey.as_hash(),
                              sig=sig)
        t = Transaction(inputs=[ti], outputs=[to])
        t.output_hash = t.calculate_output_hash()
        t.txid = t.calculate_txid()
        t.validate(self.q)

        block.transactions = [t]
        block.merkle_root = block.calculate_merkle_root()
        block.hash = block.calculate_hash()
        print("Mining Genesis Block...")
        block.ensure_difficulty(self.blockdb.pow_difficulty)

        print("Writing Genesis Block...")
        self.blockdb.write_new_block(block, self.q, genesis=True)

        # Force another consistency check to read indexes etc.
        self.blockdb.consistency_check(self.q)
        return

File: rs/src/blockchain.rs
/// Top-level blockchain orchestrator — direct port of lib/blockchain.py.
use crate::block::Block;
use crate::blockdb::BlockDB;
use crate::crypto::{ECDSAPrivateKey, Hash, Hasher};
use crate::errors::{BlockChainError, Result};
use crate::querylayer;
use crate::querylayer::QueryLayer;
use crate::transaction::{Transaction, TransactionInput, TransactionOutput};
use crate::transdb::TransDB;

pub const TOTAL_COINS: u64 = 987_654_321_000;

pub struct BlockChain {
    pub base_dir: String,
    pub blockdb: BlockDB,
    pub transdb: TransDB,
}

impl BlockChain {
    pub fn new(base_dir: &str) -> Result<Self> {
        let blockdb = BlockDB::new(base_dir)?;
        let transdb = TransDB::new(base_dir)?;

        let mut chain = BlockChain {
            base_dir: base_dir.to_string(),
            blockdb,
            transdb,
        };

        // Run consistency check
        chain.blockdb.consistency_check()?;

        Ok(chain)
    }

    pub fn get_num_blocks(&self) -> u64 {
        self.blockdb.num_blocks
    }

    pub fn get_block(&mut self, index: u64) -> Result<Block> {
        self.blockdb.get_block(index)
    }

    pub fn get_transaction(&mut self, txid: &Hash) -> Result<Transaction> {
        self.blockdb.get_transaction(txid)
    }

    pub fn validate_and_write_block(&mut self, block: &Block) -> Result<()> {
        // Validate first
        let pow = self.blockdb.pow_difficulty;
        {
            let mut q = QueryLayer::new(&mut self.blockdb);
            block.validate(&mut q, pow)?;
        }
        // Then write
        self.blockdb.write_new_block(block, false)?;
        Ok(())
    }

    pub fn write_genesis_block(&mut self, block: &Block) -> Result<()> {
        self.blockdb.write_new_block(block, true)
    }

    pub fn delete_pending_transaction(&mut self, txid: &Hash) -> Result<()> {
        self.transdb.delete_transaction(txid)
    }

    pub fn add_pending_transaction(&mut self, tx: &Transaction) -> Result<()> {
        self.transdb.add_transaction(tx)
    }

    pub fn validate_transaction(&mut self, tx: &Transaction) -> Result<()> {
        querylayer::validate_transaction(&mut self.blockdb, tx)
    }

    pub fn get_pending_transactions(&mut self) -> Result<Vec<Transaction>> {
        querylayer::get_pending_transactions(&mut self.blockdb, &mut self.transdb)
    }

    pub fn get_txids_for_address(
        &mut self,
        pubkey: &crate::crypto::ECDSAPublicKey,
        sig: &Hash,
    ) -> Result<Vec<Hash>> {
        querylayer::get_txids_for_address(&mut self.blockdb, pubkey, sig)
    }

    pub fn get_utxo(&mut self, txid: &Hash, address: &crate::crypto::CryptoAddress) -> Result<u64> {
        querylayer::get_utxo(&mut self.blockdb, txid, address)
    }

    pub fn get_utxo_private(
        &mut self,
        pubkey: &crate::crypto::ECDSAPublicKey,
        txid: &Hash,
        sig: &Hash,
    ) -> Result<u64> {
        querylayer::get_utxo_private(&mut self.blockdb, pubkey, txid, sig)
    }

    pub fn create(&mut self, passphrase: &str) -> Result<()> {
        if self.blockdb.num_blocks != 0 {
            return Err(BlockChainError::BlockValidation(
                "Blockchain already exists!".into(),
            ));
        }

        println!("Creating new blockchain...");

        // Generate genesis keys
        let pk = ECDSAPrivateKey::generate();
        let key_prefix = format!("{}/genesis_key", self.base_dir);
        pk.write_key_pair(&key_prefix, passphrase)?;

        // Build genesis block
        let mut block =
            Block::new(0, Hasher::new_with_message(b"There can be only one").get_hash())?;
        let address = pk.publickey().get_address();
        let to = TransactionOutput {
            address: address.clone(),
            amount: TOTAL_COINS,
        };

        // Fake the UTXO in
        let txid = Hasher::new_with_message(b"Genesis").get_hash();
        let utxokey = self.blockdb.get_key_for_utxo(&txid, &address);
        self.blockdb
            .db
            .put(&utxokey, &TOTAL_COINS.to_le_bytes())
            .map_err(|e| BlockChainError::DbIntegrity(format!("DB write failed: {e}")))?;

        // Sign the input
        let output_hash = Hasher::new_with_items(&[(&to.get_hash()).into()]).get_hash();
        let sig = Hasher::new_with_items(&[(&txid).into(), (&output_hash).into()]).sign(&pk)?;

        let ti = TransactionInput {
            txid: txid.clone(),
            pubkey: pk.publickey().as_hash(),
            sig,
        };

        let mut t = Transaction::new(vec![ti], vec![to])?;
        t.output_hash = t.calculate_output_hash();
        t.txid = t.calculate_txid();

        // Validate
        {
            let mut q = QueryLayer::new(&mut self.blockdb);
            t.validate(&mut q)?;
        }

        block.transactions = vec![t];
        block.merkle_root = block.calculate_merkle_root();
        block.hash = block.calculate_hash();

        println!("Mining Genesis Block...");
        block.ensure_difficulty(self.blockdb.pow_difficulty);

        println!("Writing Genesis Block...");
        self.write_genesis_block(&block)?;

        // Force another consistency check
        self.blockdb.consistency_check()?;
        Ok(())
    }
}
